libs/wingc/src/type_check.rs

```diff
@@ -4918,18 +4928,14 @@ impl<'a> TypeChecker<'a> {
 					// We can't get here twice, we can safely assume that if we're here the `object` part of the reference doesn't have and evaluated type yet.
 					// Create a type reference out of this nested reference and call ourselves again
 					let new_ref = Reference::TypeMember {
-						type_name: user_type_annotation,
+						type_name: user_type_annotation.clone(),
 						property: property.clone(),
 					};
-					// Replace the reference with the new one, this is unsafe because `reference` isn't mutable and theoretically someone may
-					// hold another reference to it. But our AST doesn't hold up/cross references so this is safe as long as we return right.
-					let const_ptr = reference as *const Reference;
-					let mut_ptr = const_ptr as *mut Reference;
-					unsafe {
-						// We don't use the return value but need to call replace so it'll drop the old value
-						_ = std::mem::replace(&mut *mut_ptr, new_ref);
-					}
-					return self.resolve_reference(reference, env);
+
+					// Store this reference for later when we can modify the final AST and replace the original reference with the new one
+					self.types.type_expressions.insert(object.id, new_ref.clone());
+
+					return self.resolve_reference(&new_ref, env);
```

Comment:
ℹ️ This is the main change required in this PR to update rust.

The mem swapping was causing the reference data to be moved, but only in release builds (not only wasm32-wasi).

This replacement now happens after type-checking is done via `type_expressions`